nevergrad/parametrization/core.py

```diff
@@ -0,0 +1,497 @@
+        if data.size:
+            raise ValueError(f"Constant dimension should be 0 (got data: {data})")
+        return instance
+
+    def spawn_child(self: P) -> P:
+        return self  # no need to create another instance for a constant
+
+
+def as_parameter(param: t.Any) -> Parameter:
+    """Returns a Parameter from anything:
+    either the input if it is already a parameter, or a Constant if not
+    This is convenient for iterating over Parameter and other objects alike
+    """
+    if isinstance(param, Parameter):
+        return param
+    else:
+        return Constant(param)
+
+
+class Dict(Parameter):
+    """Dictionary-valued parameter. This Parameter can contain other Parameters,
+    its value is a dict, with keys the ones provided as input, and corresponding values are
+    either directly the provided values if they are not Parameter instances, or the value of those
+    Parameters. It also implements a getter to access the Parameters directly if need be.
+
+    Parameters
+    ----------
+    **parameters: Any
+        the objects or Parameter which will provide values for the dict
+
+    Note
+    ----
+    This is the base structure for all container Parameters, and it is
+    used to hold the subparameters for all Parameter classes.
+    """
+
+    def __init__(self, **parameters: t.Any) -> None:
+        super().__init__()
+        self._parameters: t.Dict[t.Any, t.Any] = parameters
+        self._sizes: t.Optional[t.Dict[str, int]] = None
+
+    @property
+    def descriptors(self) -> Descriptors:
+        return Descriptors(**{name: all(getattr(as_parameter(p).descriptors, name) for p in self._parameters.values())
+                              for name in ("deterministic", "continuous")})
+
+    def __getitem__(self, name: t.Any) -> t.Any:
+        return self._parameters[name]
+
+    def __len__(self) -> int:
+        return len(self._parameters)
+
+    def _get_parameters_str(self) -> str:
+        params = sorted((k, as_parameter(p).name) for k, p in self._parameters.items())
+        return ",".join(f"{k}={n}" for k, n in params)
+
+    def _get_name(self) -> str:
+        return f"{self.__class__.__name__}({self._get_parameters_str()})"
+
+    @property
+    def value(self) -> t.Dict[str, t.Any]:
+        return {k: as_parameter(p).value for k, p in self._parameters.items()}
+
+    @value.setter
+    def value(self, value: t.Dict[str, t.Any]) -> None:
+        if set(value) != set(self._parameters):
+            raise ValueError(f"Got input keys {set(value)} but expected {set(self._parameters)}")
+        for key, val in value.items():
+            as_parameter(self._parameters[key]).value = val
+
+    def get_value_hash(self) -> t.Hashable:
+        return tuple(sorted((x, y.get_value_hash()) for x, y in self._parameters.items() if isinstance(y, Parameter)))
+
+    def _internal_get_std_data(self: D, instance: D) -> np.ndarray:
```

Comment:
nit: do you need to type `self`?

Reply:
yes, because `instance` must be of the same type as `self` 